crates/uv-distribution/src/source/mod.rs

```diff
@@ -1215,3 +1159,162 @@ fn read_wheel_metadata(
+        Ok(metadata) => metadata,
+        Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+            return Err(Error::NotFound(path.to_path_buf()));
+        }
+        Err(err) => return Err(Error::CacheRead(err)),
+    };
+
+    if metadata.is_dir() {
+        Ok(ExtractedSource::Directory(path.to_path_buf()))
+    } else {
+        debug!("Unpacking for build: {}", path.display());
+
+        let temp_dir = tempfile::tempdir_in(cache.bucket(CacheBucket::BuiltWheels))
+            .map_err(Error::CacheWrite)?;
+
+        // Unzip the archive into the temporary directory.
+        let reader = fs_err::tokio::File::open(&path)
+            .await
+            .map_err(Error::CacheRead)?;
+        uv_extract::seek::archive(tokio::io::BufReader::new(reader), path, &temp_dir.path())
+            .await?;
+
+        // Extract the top-level directory from the archive.
+        let extracted = match uv_extract::strip_component(temp_dir.path()) {
+            Ok(top_level) => top_level,
+            Err(uv_extract::Error::NonSingularArchive(_)) => temp_dir.path().to_path_buf(),
+            Err(err) => return Err(err.into()),
+        };
+
+        Ok(ExtractedSource::Archive(extracted, temp_dir))
+    }
+}
+
+/// Download a source distribution from a Git repository.
+async fn fetch_git_archive(
+    url: &Url,
+    cache: &Cache,
+    reporter: Option<&Arc<dyn Reporter>>,
+) -> Result<(Fetch, Option<PathBuf>), Error> {
+    debug!("Fetching source distribution from Git: {url}");
+    let git_dir = cache.bucket(CacheBucket::Git);
+
+    // Avoid races between different processes, too.
+    let lock_dir = git_dir.join("locks");
+    fs::create_dir_all(&lock_dir)
+        .await
+        .map_err(Error::CacheWrite)?;
+    let canonical_url = cache_key::CanonicalUrl::new(url);
+    let _lock = LockedFile::acquire(
+        lock_dir.join(cache_key::digest(&canonical_url)),
+        &canonical_url,
+    )
+    .map_err(Error::CacheWrite)?;
+
+    let DirectGitUrl { url, subdirectory } = DirectGitUrl::try_from(url).map_err(Error::Git)?;
+
+    let source = if let Some(reporter) = reporter {
+        GitSource::new(url, git_dir).with_reporter(Facade::from(reporter.clone()))
+    } else {
+        GitSource::new(url, git_dir)
+    };
+    let fetch = tokio::task::spawn_blocking(move || source.fetch())
+        .await?
+        .map_err(Error::Git)?;
+    Ok((fetch, subdirectory))
+}
+
+/// Download and extract a source distribution from a URL.
+///
+/// This function will download the source distribution from the given URL, and extract it into a
+/// directory.
+///
+/// For VCS distributions, this method will checkout the URL into the shared Git cache.
+///
+/// For local archives, this method will extract the archive into a temporary directory.
+///
+/// For HTTP distributions, this method will download the archive and extract it into a temporary
+/// directory.
+pub async fn download_archive(
```

Comment:
Does it feel worth clarifying that extraction occurs in the name here? It was not obvious until I pulled up an IDE to look at the type signature.